tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -783,6 +803,52 @@ export class WebGPUBackend extends KernelBackend {
     this.uploadToGPU(output.dataId);
     program.dispatch = reshapeDispatch(this.device, program);
 
+    const tensorShapes: number[][] = inputs.concat(output).map(d => d.shape);
+
+    const inputsData = inputs.map((input: TensorInfo, i: number) => {
+      if (input.dtype === 'complex64') {
+        throw new Error(
+            `GPGPUProgram does not support complex64 input. For complex64 ` +
+            `dtypes, please separate the program into real and imaginary ` +
+            `parts.`);
+      }
+      this.uploadToGPU(input.dataId);
+
+      return {
+        // Returning dtype from tensorMap because it reflects dtype
+        // of underlying buffer, rather than abstract dtype.
+        dtype: this.tensorMap.get(input.dataId).dtype,
+        shape: input.shape,
+        name: program.variableNames[i]
+      };
+    });
+
+    program.shaderKey =
+        webgpu_program.makeShaderKey(program, tensorShapes, inputsData, output);
```

Comment:
Since the program.shaderKey has been assigned to the full shader key, do you still need to pass `program.shaderKey` to `compileProgram in L831?

Reply:
It seems that `tensorShapes` can also be removed since you have `inputsData` and `output`.

Reply:
Done